File: src/iambic.rs
//! Iambic keyer finite state machine.
//!
//! Pure logic, no hardware dependencies. Consumes GPIO state,
//! produces keying output. Fully testable on host.
//!
//! # Iambic Modes
//!
//! - **Mode A**: Stops immediately when paddles released
//! - **Mode B**: Completes current element + one more when squeeze released

use crate::sample::{GpioState, StreamSample};

/// Iambic keyer mode.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum IambicMode {
    /// Mode A: Stop immediately when paddles released.
    A,
    /// Mode B: Complete current + one bonus element on squeeze release.
    B,
}

impl Default for IambicMode {
    fn default() -> Self {
        IambicMode::B
    }
}

/// Keying element type.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Element {
    Dit,
    Dah,
}

impl Element {
    /// Get the opposite element.
    #[inline]
    pub fn opposite(self) -> Self {
        match self {
            Element::Dit => Element::Dah,
            Element::Dah => Element::Dit,
        }
    }
}

/// FSM state.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
enum State {
    Idle,
    SendDit,
    SendDah,
    Gap,
}

/// Iambic keyer configuration.
#[derive(Clone, Copy, Debug)]
pub struct IambicConfig {
    /// Speed in words per minute (PARIS timing).
    pub wpm: u32,

    /// Iambic mode (A or B).
    pub mode: IambicMode,

    /// Dit memory enable.
    pub dit_memory: bool,

    /// Dah memory enable.
    pub dah_memory: bool,
}

impl Default for IambicConfig {
    fn default() -> Self {
        Self {
            wpm: 20,
            mode: IambicMode::B,
            dit_memory: true,
            dah_memory: true,
        }
    }
}

impl IambicConfig {
    /// Create config for given WPM with default settings.
    pub fn with_wpm(wpm: u32) -> Self {
        Self {
            wpm,
            ..Default::default()
        }
    }

    /// Calculate dit duration in microseconds.
    ///
    /// PARIS timing: dit = 1.2 / WPM seconds
    #[inline]
    pub fn dit_duration_us(&self) -> i64 {
        1_200_000 / self.wpm as i64
    }

    /// Calculate dah duration in microseconds (3x dit).
    #[inline]
    pub fn dah_duration_us(&self) -> i64 {
        self.dit_duration_us() * 3
    }

    /// Calculate inter-element gap in microseconds (1x dit).
    #[inline]
    pub fn gap_duration_us(&self) -> i64 {
        self.dit_duration_us()
    }
}

/// Iambic keyer processor.
///
/// Pure FSM that converts paddle GPIO state into keying output.
/// No hardware dependencies, fully testable on host.
///
/// # Example
///
/// ```
/// use rust_remote_cw_keyer::iambic::{IambicProcessor, IambicConfig};
/// use rust_remote_cw_keyer::sample::GpioState;
///
/// let mut keyer = IambicProcessor::new(IambicConfig::with_wpm(25));
///
/// // Simulate DIT paddle press
/// let mut gpio = GpioState::new();
/// gpio.set_dit(true);
///
/// let sample = keyer.tick(0, gpio);
/// assert!(sample.local_key); // Key down
/// ```
pub struct IambicProcessor {
    config: IambicConfig,

    // FSM state
    state: State,
    element_end_us: i64,
    last_element: Element,

    // Paddle state
    dit_pressed: bool,
    dah_pressed: bool,

    // Memory flags
    dit_memory: bool,
    dah_memory: bool,

    // Mode B: squeeze tracking
    squeeze_seen: bool,

    // Output state
    key_down: bool,
}

impl IambicProcessor {
    /// Create a new iambic processor with given configuration.
    pub fn new(config: IambicConfig) -> Self {
        Self {
            config,
            state: State::Idle,
            element_end_us: 0,
            last_element: Element::Dah, // Start with Dah so first press of Dit works
            dit_pressed: false,
            dah_pressed: false,
            dit_memory: false,
            dah_memory: false,
            squeeze_seen: false,
            key_down: false,
        }
    }

    /// Update configuration (e.g., change speed).
    pub fn set_config(&mut self, config: IambicConfig) {
        self.config = config;
    }

    /// Get current configuration.
    pub fn config(&self) -> &IambicConfig {
        &self.config
    }

    /// Tick the FSM and produce output sample.
    ///
    /// # Arguments
    ///
    /// * `now_us` - Current timestamp in microseconds
    /// * `gpio` - Current paddle GPIO state
    ///
    /// # Returns
    ///
    /// StreamSample with `local_key` set to current keying state.
    #[inline]
    pub fn tick(&mut self, now_us: i64, gpio: GpioState) -> StreamSample {
        // Update paddle state
        self.update_gpio(gpio);

        // Run FSM
        match self.state {
            State::Idle => self.tick_idle(now_us),
            State::SendDit => self.tick_sending(now_us, Element::Dit),
            State::SendDah => self.tick_sending(now_us, Element::Dah),
            State::Gap => self.tick_gap(now_us),
        }

        // Produce output sample
        StreamSample {
            gpio,
            local_key: self.key_down,
            audio_level: 0,  // Filled by audio envelope later
            flags: 0,
            config_gen: 0,   // Set by RT loop
        }
    }

    /// Check if key output is currently active.
    #[inline]
    pub fn is_key_down(&self) -> bool {
        self.key_down
    }

    /// Reset FSM to idle state.
    pub fn reset(&mut self) {
        self.state = State::Idle;
        self.element_end_us = 0;
        self.dit_memory = false;
        self.dah_memory = false;
        self.squeeze_seen = false;
        self.key_down = false;
    }

    // --- Private methods ---

    fn update_gpio(&mut self, gpio: GpioState) {
        let was_squeeze = self.dit_pressed && self.dah_pressed;

        self.dit_pressed = gpio.dit();
        self.dah_pressed = gpio.dah();

        let is_squeeze = self.dit_pressed && self.dah_pressed;

        // Track squeeze for Mode B
        if is_squeeze && !was_squeeze {
            self.squeeze_seen = true;
        }

        // Arm memory when paddle pressed during element/gap
        if self.state != State::Idle {
            if self.dit_pressed && self.config.dit_memory {
                self.dit_memory = true;
            }
            if self.dah_pressed && self.config.dah_memory {
                self.dah_memory = true;
            }
        }
    }

    fn tick_idle(&mut self, now_us: i64) {
        // Determine next element from memory or current paddle state
        let next = self.decide_next_element();

        if let Some(element) = next {
            self.start_element(element, now_us);
        }
    }

    fn tick_sending(&mut self, now_us: i64, element: Element) {
        if now_us >= self.element_end_us {
            // Element complete
            self.key_down = false;
            self.last_element = element;

            // Enter gap
            self.state = State::Gap;
            self.element_end_us = now_us + self.config.gap_duration_us();
        }
    }

    fn tick_gap(&mut self, now_us: i64) {
        if now_us >= self.element_end_us {
            // Gap complete
            self.state = State::Idle;

            // Immediately check for next element
            self.tick_idle(now_us);
        }
    }

    fn decide_next_element(&mut self) -> Option<Element> {
        // Priority 1: Memory (armed during previous element)
        if self.dit_memory {
            self.dit_memory = false;
            return Some(Element::Dit);
        }
        if self.dah_memory {
            self.dah_memory = false;
            return Some(Element::Dah);
        }

        // Priority 2: Mode B bonus element
        if self.config.mode == IambicMode::B && self.squeeze_seen {
            // Check if squeeze was released
            if !(self.dit_pressed && self.dah_pressed) {
                self.squeeze_seen = false;
                return Some(self.last_element.opposite());
            }
        }

        // Priority 3: Current paddle state
        match (self.dit_pressed, self.dah_pressed) {
            (true, true) => {
                // Squeeze: alternate from last element
                Some(self.last_element.opposite())
            }
            (true, false) => Some(Element::Dit),
            (false, true) => Some(Element::Dah),
            (false, false) => {
                // No paddle pressed
                self.squeeze_seen = false;
                None
            }
        }
    }

    fn start_element(&mut self, element: Element, now_us: i64) {
        self.key_down = true;
        self.squeeze_seen = self.dit_pressed && self.dah_pressed;

        let duration = match element {
            Element::Dit => self.config.dit_duration_us(),
            Element::Dah => self.config.dah_duration_us(),
        };

        self.state = match element {
            Element::Dit => State::SendDit,
            Element::Dah => State::SendDah,
        };

        self.element_end_us = now_us + duration;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_single_dit() {
        let config = IambicConfig::with_wpm(20);
        let dit_us = config.dit_duration_us();
        let gap_us = config.gap_duration_us();

        let mut keyer = IambicProcessor::new(config);

        // Press DIT
        let mut gpio = GpioState::new();
        gpio.set_dit(true);

        // Tick 0: should start dit
        let sample = keyer.tick(0, gpio);
        assert!(sample.local_key);

        // Mid-dit: still on
        let sample = keyer.tick(dit_us / 2, gpio);
        assert!(sample.local_key);

        // Dit complete: key off, in gap
        let sample = keyer.tick(dit_us, gpio);
        assert!(!sample.local_key);

        // Release paddle during gap
        gpio.set_dit(false);
        let sample = keyer.tick(dit_us + gap_us / 2, gpio);
        assert!(!sample.local_key);

        // Gap complete, no paddle: should stay idle
        let sample = keyer.tick(dit_us + gap_us + 1000, gpio);
        assert!(!sample.local_key);
    }

    #[test]
    fn test_squeeze_alternates() {
        let config = IambicConfig::with_wpm(20);
        let dit_us = config.dit_duration_us();
        let dah_us = config.dah_duration_us();
        let gap_us = config.gap_duration_us();

        let mut keyer = IambicProcessor::new(config);

        // Press both (squeeze)
        let mut gpio = GpioState::new();
        gpio.set_dit(true);
        gpio.set_dah(true);

        // Should start with Dit (opposite of initial last_element=Dah)
        let sample = keyer.tick(0, gpio);
        assert!(sample.local_key);

        // Complete dit + gap
        let t = dit_us + gap_us;
        let sample = keyer.tick(t, gpio);

        // Should now be sending Dah
        assert!(sample.local_key);

        // Complete dah + gap
        let t = t + dah_us + gap_us;
        let sample = keyer.tick(t, gpio);

        // Should be sending Dit again
        assert!(sample.local_key);
    }

    #[test]
    fn test_mode_b_bonus_element() {
        let config = IambicConfig {
            wpm: 20,
            mode: IambicMode::B,
            ..Default::default()
        };
        let dit_us = config.dit_duration_us();
        let gap_us = config.gap_duration_us();

        let mut keyer = IambicProcessor::new(config);

        // Press both (squeeze)
        let mut gpio = GpioState::new();
        gpio.set_dit(true);
        gpio.set_dah(true);

        // Start dit
        keyer.tick(0, gpio);

        // Release during dit
        gpio.set_dit(false);
        gpio.set_dah(false);
        keyer.tick(dit_us / 2, gpio);

        // Complete dit + gap
        let t = dit_us + gap_us;
        let sample = keyer.tick(t, gpio);

        // Mode B: should send bonus dah even though released
        assert!(sample.local_key);
    }

    #[test]
    fn test_dit_timing_accuracy() {
        let config = IambicConfig::with_wpm(25);
        let expected_dit_us = 48_000; // 1.2s / 25 = 48ms

        assert_eq!(config.dit_duration_us(), expected_dit_us);
    }

    #[test]
    fn test_memory() {
        let config = IambicConfig::with_wpm(20);
        let dit_us = config.dit_duration_us();
        let dah_us = config.dah_duration_us();
        let gap_us = config.gap_duration_us();

        let mut keyer = IambicProcessor::new(config);

        // Start with dah only
        let mut gpio = GpioState::new();
        gpio.set_dah(true);

        keyer.tick(0, gpio);
        assert!(keyer.is_key_down());

        // During dah, tap dit (memory)
        gpio.set_dit(true);
        keyer.tick(dah_us / 2, gpio);
        gpio.set_dit(false);
        keyer.tick(dah_us / 2 + 1000, gpio);

        // Release dah
        gpio.set_dah(false);
        keyer.tick(dah_us - 1000, gpio);

        // Complete dah + gap
        let t = dah_us + gap_us;
        let sample = keyer.tick(t, gpio);

        // Should send memorized dit
        assert!(sample.local_key);
    }
}

File: components/keyer_iambic/src/iambic.c
/**
 * @file iambic.c
 * @brief Iambic keyer FSM implementation
 *
 * Pure logic, no I/O or allocation. Fully testable on host.
 *
 * Memory Window Logic:
 * Paddle inputs are only memorized when within the memory window
 * (between mem_window_start_pct and mem_window_end_pct of element duration).
 */

#include "iambic.h"
#include <assert.h>
#include <string.h>

/* ============================================================================
 * Forward Declarations
 * ============================================================================ */

static void update_gpio(iambic_processor_t *proc, gpio_state_t gpio, int64_t now_us);
static void tick_idle(iambic_processor_t *proc, int64_t now_us);
static void tick_sending(iambic_processor_t *proc, int64_t now_us, iambic_element_t element);
static void tick_gap(iambic_processor_t *proc, int64_t now_us);
static iambic_element_t *decide_next_element(iambic_processor_t *proc, iambic_element_t *out);
static void start_element(iambic_processor_t *proc, iambic_element_t element, int64_t now_us);
static bool is_in_memory_window(const iambic_processor_t *proc, int64_t now_us);

/* ============================================================================
 * Public Functions
 * ============================================================================ */

void iambic_init(iambic_processor_t *proc, const iambic_config_t *config) {
    assert(proc != NULL);
    assert(config != NULL);

    proc->config = *config;
    proc->state = IAMBIC_STATE_IDLE;
    proc->element_start_us = 0;
    proc->element_end_us = 0;
    proc->element_duration_us = 0;
    proc->last_element = ELEMENT_DAH;  /* Start with DAH so first DIT press works */
    proc->dit_pressed = false;
    proc->dah_pressed = false;
    proc->dit_release_time_us = 0;
    proc->dah_release_time_us = 0;
    proc->dit_press_start_us = 0;
    proc->dah_press_start_us = 0;
    proc->dit_memory = false;
    proc->dah_memory = false;
    proc->squeeze_seen = false;
    proc->squeeze_latched = false;
    proc->key_down = false;
}

void iambic_set_config(iambic_processor_t *proc, const iambic_config_t *config) {
    assert(proc != NULL);
    assert(config != NULL);

    proc->config = *config;
}

stream_sample_t iambic_tick(iambic_processor_t *proc, int64_t now_us, gpio_state_t gpio) {
    assert(proc != NULL);

    /* Update paddle state and memory */
    update_gpio(proc, gpio, now_us);

    /* Run FSM */
    switch (proc->state) {
        case IAMBIC_STATE_IDLE:
            tick_idle(proc, now_us);
            break;
        case IAMBIC_STATE_SEND_DIT:
            tick_sending(proc, now_us, ELEMENT_DIT);
            break;
        case IAMBIC_STATE_SEND_DAH:
            tick_sending(proc, now_us, ELEMENT_DAH);
            break;
        case IAMBIC_STATE_GAP:
            tick_gap(proc, now_us);
            break;
    }

    /* Produce output sample */
    stream_sample_t sample = STREAM_SAMPLE_EMPTY;
    sample.gpio = gpio;
    sample.local_key = proc->key_down ? 1 : 0;
    /* audio_level filled by audio envelope later */
    /* flags and config_gen set by RT loop */

    return sample;
}

void iambic_reset(iambic_processor_t *proc) {
    assert(proc != NULL);

    proc->state = IAMBIC_STATE_IDLE;
    proc->element_start_us = 0;
    proc->element_end_us = 0;
    proc->element_duration_us = 0;
    proc->dit_memory = false;
    proc->dah_memory = false;
    proc->squeeze_seen = false;
    proc->squeeze_latched = false;
    proc->key_down = false;
    proc->dit_release_time_us = 0;
    proc->dah_release_time_us = 0;
    proc->dit_press_start_us = 0;
    proc->dah_press_start_us = 0;
}

/* ============================================================================
 * Private Functions
 * ============================================================================ */

/**
 * @brief Check if current time is within the memory window of current element
 */
static bool is_in_memory_window(const iambic_processor_t *proc, int64_t now_us) {
    /* Only applies during element transmission (not gap) */
    if (proc->state != IAMBIC_STATE_SEND_DIT && proc->state != IAMBIC_STATE_SEND_DAH) {
        return false;  /* NEVER arm memory during gap - use current paddle state */
    }

    if (proc->element_duration_us <= 0) {
        return true;  /* Fallback: no window defined */
    }

    /* Calculate progress percentage (0-100) */
    int64_t elapsed = now_us - proc->element_start_us;
    if (elapsed < 0) {
        elapsed = 0;
    }

    /* Avoid overflow: (elapsed * 100) could overflow for large durations */
    uint8_t progress_pct;
    if (elapsed >= proc->element_duration_us) {
        progress_pct = 100;
    } else {
        progress_pct = (uint8_t)((elapsed * 100) / proc->element_duration_us);
    }

    return iambic_in_memory_window(progress_pct,
                                    proc->config.mem_window_start_pct,
                                    proc->config.mem_window_end_pct);
}

static void update_gpio(iambic_processor_t *proc, gpio_state_t gpio, int64_t now_us) {
    bool was_dit_pressed = proc->dit_pressed;
    bool was_dah_pressed = proc->dah_pressed;
    bool was_squeeze = was_dit_pressed && was_dah_pressed;

    /* Read raw GPIO state */
    bool raw_dit = gpio_dit(gpio);
    bool raw_dah = gpio_dah(gpio);

    /* Detect release transitions and record timestamps for debounce */
    if (was_dit_pressed && !raw_dit) {
        proc->dit_release_time_us = now_us;
    }
    if (was_dah_pressed && !raw_dah) {
        proc->dah_release_time_us = now_us;
    }

    /* Apply release debounce: ignore press if within blanking period after release.
     * This suppresses false triggers from mechanical contact bounce on opening. */
    bool dit_in_blanking = (now_us - proc->dit_release_time_us) < IAMBIC_DEBOUNCE_RELEASE_US;
    bool dah_in_blanking = (now_us - proc->dah_release_time_us) < IAMBIC_DEBOUNCE_RELEASE_US;

    bool new_dit_pressed = raw_dit && !dit_in_blanking;
    bool new_dah_pressed = raw_dah && !dah_in_blanking;

    /* Track rising edge (press start) for fresh press detection */
    if (new_dit_pressed && !proc->dit_pressed) {
        proc->dit_press_start_us = now_us;
    }
    if (new_dah_pressed && !proc->dah_pressed) {
        proc->dah_press_start_us = now_us;
    }

    proc->dit_pressed = new_dit_pressed;
    proc->dah_pressed = new_dah_pressed;

    bool is_squeeze = proc->dit_pressed && proc->dah_pressed;

    /* Track squeeze for Mode B */
    if (is_squeeze && !was_squeeze) {
        proc->squeeze_seen = true;
    }

    /* Handle squeeze mode (LATCH_ON vs LATCH_OFF) */
    if (proc->config.squeeze_mode == SQUEEZE_MODE_LATCH_ON) {
        /* LATCH_ON: Capture squeeze state at element start, don't update during element */
        /* squeeze_latched is set in start_element() and used for squeeze detection */
    } else {
        /* LATCH_OFF: Live mode - continuously update squeeze_latched */
        proc->squeeze_latched = is_squeeze;
    }

    /* Arm memory ONLY during element transmission (not gap, not idle)
     * Memory window prevents arming at beginning (debounce) and end (too late).
     * During gap, we rely on Priority 3 (current paddle state) for squeeze alternation. */
    if (proc->state == IAMBIC_STATE_SEND_DIT || proc->state == IAMBIC_STATE_SEND_DAH) {
        bool in_window = is_in_memory_window(proc, now_us);
        bool can_arm_dit = (proc->state != IAMBIC_STATE_SEND_DIT);
        bool can_arm_dah = (proc->state != IAMBIC_STATE_SEND_DAH);

        /* Determine if we should check for memory arming based on squeeze_mode */
        bool check_dit, check_dah;
        if (proc->config.squeeze_mode == SQUEEZE_MODE_LATCH_ON) {
            /* LATCH_ON: Only arm memory if squeeze was active at element start */
            check_dit = proc->squeeze_latched && proc->dit_pressed;
            check_dah = proc->squeeze_latched && proc->dah_pressed;
        } else {
            /* LATCH_OFF: Live mode - check current paddle state */
            check_dit = proc->dit_pressed;
            check_dah = proc->dah_pressed;
        }

        if (in_window) {
            /* Fresh press detection: only arm memory if press started AFTER element began.
             * This prevents re-arming from the same press due to contact bounce. */
            bool dit_is_fresh = (proc->dit_press_start_us > proc->element_start_us);
            bool dah_is_fresh = (proc->dah_press_start_us > proc->element_start_us);

            /* Inside window: arm memory if paddle pressed AND it's a fresh press */
            if (can_arm_dit && check_dit && dit_is_fresh && iambic_dit_memory_enabled(proc->config.memory_mode)) {
                proc->dit_memory = true;
            }
            if (can_arm_dah && check_dah && dah_is_fresh && iambic_dah_memory_enabled(proc->config.memory_mode)) {
                proc->dah_memory = true;
            }
        }

        /* Reset squeeze_seen if squeeze released BEFORE memory window (Mode B fix) */
        if (proc->config.mode == IAMBIC_MODE_B && proc->squeeze_seen) {
            bool current_squeeze = (proc->config.squeeze_mode == SQUEEZE_MODE_LATCH_ON)
                                    ? proc->squeeze_latched
                                    : (proc->dit_pressed && proc->dah_pressed);
            if (!current_squeeze) {
                /* Squeeze released - check if before window */
                int64_t elapsed = now_us - proc->element_start_us;
                uint8_t progress_pct = (proc->element_duration_us > 0)
                    ? (uint8_t)((elapsed * 100) / proc->element_duration_us)
                    : 0;

                if (progress_pct < proc->config.mem_window_start_pct) {
                    /* Released BEFORE window start - cancel Mode B bonus */
                    proc->squeeze_seen = false;
                }
            }
        }
    }
}

static void tick_idle(iambic_processor_t *proc, int64_t now_us) {
    /* Determine next element from memory or current paddle state */
    iambic_element_t next_element;
    const iambic_element_t *next = decide_next_element(proc, &next_element);

    if (next != NULL) {
        start_element(proc, *next, now_us);
    }
}

static void tick_sending(iambic_processor_t *proc, int64_t now_us, iambic_element_t element) {
    if (now_us >= proc->element_end_us) {
        /* Element complete */
        proc->key_down = false;
        proc->last_element = element;

        /* Enter gap */
        proc->state = IAMBIC_STATE_GAP;
        proc->element_start_us = now_us;
        proc->element_duration_us = iambic_gap_duration_us(&proc->config);
        proc->element_end_us = now_us + proc->element_duration_us;
    }
}

static void tick_gap(iambic_processor_t *proc, int64_t now_us) {
    if (now_us >= proc->element_end_us) {
        /* Gap complete */
        proc->state = IAMBIC_STATE_IDLE;
        proc->element_duration_us = 0;

        /* Immediately check for next element */
        tick_idle(proc, now_us);
    }
}

static iambic_element_t *decide_next_element(iambic_processor_t *proc, iambic_element_t *out) {
    /* Priority 1: Memory (armed during previous element) */
    if (proc->dit_memory) {
        proc->dit_memory = false;
        *out = ELEMENT_DIT;
        return out;
    }
    if (proc->dah_memory) {
        proc->dah_memory = false;
        *out = ELEMENT_DAH;
        return out;
    }

    /* Priority 2: Mode B bonus element */
    if (proc->config.mode == IAMBIC_MODE_B && proc->squeeze_seen) {
        /* Check if squeeze was released */
        bool current_squeeze = (proc->config.squeeze_mode == SQUEEZE_MODE_LATCH_ON)
                                ? proc->squeeze_latched
                                : (proc->dit_pressed && proc->dah_pressed);
        if (!current_squeeze) {
            proc->squeeze_seen = false;
            *out = iambic_element_opposite(proc->last_element);
            return out;
        }
    }

    /* Priority 3: Current paddle state */
    if (proc->dit_pressed && proc->dah_pressed) {
        /* Squeeze: alternate from last element */
        *out = iambic_element_opposite(proc->last_element);
        return out;
    } else if (proc->dit_pressed) {
        *out = ELEMENT_DIT;
        return out;
    } else if (proc->dah_pressed) {
        *out = ELEMENT_DAH;
        return out;
    } else {
        /* No paddle pressed */
        proc->squeeze_seen = false;
        return NULL;
    }
}

static void start_element(iambic_processor_t *proc, iambic_element_t element, int64_t now_us) {
    proc->key_down = true;

    /* Latch squeeze state at element start (for LATCH_ON mode) */
    proc->squeeze_latched = proc->dit_pressed && proc->dah_pressed;
    proc->squeeze_seen = proc->squeeze_latched;

    int64_t duration;
    switch (element) {
        case ELEMENT_DIT:
            duration = iambic_dit_duration_us(&proc->config);
            proc->state = IAMBIC_STATE_SEND_DIT;
            break;
        case ELEMENT_DAH:
            duration = iambic_dah_duration_us(&proc->config);
            proc->state = IAMBIC_STATE_SEND_DAH;
            break;
        default:
            /* Should not happen */
            duration = 0;
            break;
    }

    proc->element_start_us = now_us;
    proc->element_duration_us = duration;
    proc->element_end_us = now_us + duration;
}
